### PR TITLE
fix(apigatewayv2): HttpApi and Route in different stacks creates cycles

### LINKED ISSUE #13021: (apigatewayv2): cyclic dependency introduced when HttpApi and Route are in different stacks
Please add your +1 👍 to let us know you have encountered this
---

### Overview:

From @hoegertn:

> The changes in #12528 introduced a regression when having the following setup:
> 
> ApiStack -> ChildStack with Lambda, HttpRoute, etc
> 
> By moving the instantiation of the HttpIntegration class to the `HttpApi. _addIntegration` method the scope was changed from the Route to the HttpApi object. This leads to a cyclic reference in the architecture described above.
> 
> So starting with 1.87.0 all my ApiGateway applications no longer synthesize.

#### Complete Error Message:
```
'myapp-backend-pipeline/myapp-backend-dev/Api' depends on 'myapp-backend-pipeline/myapp-backend-dev/Api/Config' (myapp-backend-pipeline/myapp-backend-dev/Api -> myapp-backend-pipeline/myapp-backend-dev/Api/Config/Fncors/Resource.Arn, myapp-backend-pipeline/myapp-backend-dev/Api -> myapp-backend-pipeline/myapp-backend-dev/Api/Config/Fnopenapi/Resource.Arn). Adding this dependency (myapp-backend-pipeline/myapp-backend-dev/Api/Config -> myapp-backend-pipeline/myapp-backend-dev/Api/Api/Resource/Resource.Ref) would create a cyclic reference.
Error: 'myapp-backend-pipeline/myapp-backend-dev/Api' depends on 'myapp-backend-pipeline/myapp-backend-dev/Api/Config' (myapp-backend-pipeline/myapp-backend-dev/Api -> myapp-backend-pipeline/myapp-backend-dev/Api/Config/Fncors/Resource.Arn, myapp-backend-pipeline/myapp-backend-dev/Api -> myapp-backend-pipeline/myapp-backend-dev/Api/Config/Fnopenapi/Resource.Arn). Adding this dependency (myapp-backend-pipeline/myapp-backend-dev/Api/Config -> myapp-backend-pipeline/myapp-backend-dev/Api/Api/Resource/Resource.Ref) would create a cyclic reference.
    at BackendApiImplStack._addAssemblyDependency (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/stack.ts:724:13)
    at Object.addDependency (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/deps.ts:52:20)
    at BackendApiImplStack.addDependency (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/stack.ts:479:5)
    at resolveValue (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/private/refs.ts:103:12)
    at Object.resolveReferences (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/private/refs.ts:33:24)
    at Object.prepareApp (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/private/prepare-app.ts:31:3)
    at Object.synthesize (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/private/synthesis.ts:24:3)
    at myappBackendStage.synth (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/stage.ts:188:23)
    at synthNestedAssemblies (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/private/synthesis.ts:52:13)
    at synthNestedAssemblies (/Users/hoegertn/repo/myapp/backend/node_modules/@aws-cdk/core/lib/private/synthesis.ts:54:7)
```

### Related Issues:

First reported by @hoegertn: #13007

### PR DESCRIPTION
**Pull request by hoegertn**
When the Route construct with an integration is defined
in a separate stack from the one defining the HttpApi
construct, a circular dependency error is thrown.

The bug was introduced in - 855ce590 - where the
refactor of moving the creation of Integration resulted in
using the HttpApi as the scope of the Integration
construct, rather than the Route.

fixes #13021

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by hoegertn** [comment]
@eladb @nija-at I modified the code to keep the old semantic of using the route scope for the integration.

But I see the benefit of the Integration being tied to the API scope.

I think ideally the integration should be scoped depending on the integration type and for Lambda integrations it should be in the same scope as the function, as this causes the cyclic dependency.

**Comment by nija-at** [comment]
> But I see the benefit of the Integration being tied to the API scope.

I'm curious what the benefit is?

> I think ideally the integration should be scoped depending on the integration type and for Lambda integrations it should be in the same scope as the function, as this causes the cyclic dependency.

Do you mean when the lambda is in the API scope or just any other stack?

**Comment by hoegertn** [comment]
> > But I see the benefit of the Integration being tied to the API scope.
> 
> I'm curious what the benefit is?

The set of integrations for deduplication is located at the API and integrations are possibly reused, so they should live at the top level.

> 
> > I think ideally the integration should be scoped depending on the integration type and for Lambda integrations it should be in the same scope as the function, as this causes the cyclic dependency.
> 
> Do you mean when the lambda is in the API scope or just any other stack?
> 
> It feels reasonable to assume that the lambda function will be defined closer to the integration rather than the API.

The integration should live in the same stack as the Lambda function at any time. Just not sure about other integrations that are not Lambda. 

But I am totally happy to have integrations located at the route as in my PR, I am just not sure as I never reused integrations and want to see if this breaks other peoples workflows.

**Comment by nija-at** [comment]
> The integration should live in the same stack as the Lambda function at any time.

It could also live in another stack, no?

How about taking `scope` as an optional parameter as part of integration's props and defaulting to the `Route`'s scope if not provided?

**Comment by hoegertn** [comment]
> > The integration should live in the same stack as the Lambda function at any time.
> 
> It could also live in another stack, no?

It references the Lambda function so it could only live in the same stack or downstream but not more to the top.

Api -> Lambda -> Integ

**Comment by nija-at** [comment]
>>> The integration should live in the same stack as the Lambda function at any time.
>>
>> It could also live in another stack, no?
>
> It references the Lambda function so it could only live in the same stack or downstream but not more to the top.
>
> Api -> Lambda -> Integ

Agreed.

**Comment by ayush987goyal** [comment]
@hoegertn Thanks for this change! I did not realise this flow while making the change. 

I am just curious on why have you decided to make the `scope` as optional and defaulted to `this` in the `_addIntegration` helper? I think we should always make it scoped in the `Route`?

**Comment by nija-at** [comment]
> @hoegertn Thanks for this change! I did not realise this flow while making the change.
> 
> I am just curious on why have you decided to make the `scope` as optional and defaulted to `this` in the `_addIntegration` helper? I think we should always make it scoped in the `Route`?

Good point. And we should make `scope` the first argument on the method, just like in any Construct.

**Comment by hoegertn** [comment]
> @hoegertn Thanks for this change! I did not realise this flow while making the change.
> 
> I am just curious on why have you decided to make the `scope` as optional and defaulted to `this` in the `_addIntegration` helper? I think we should always make it scoped in the `Route`?

The `_addIntegration` could be called from another place outside of the scope of a route. So I thought a default of the API as the parent might be a good idea.

**Comment by ayush987goyal** [comment]
>The _addIntegration could be called from another place outside of the scope of a route. So I thought a default of the API as the parent might be a good idea.

I am just thinking if this problem would then occur in that context? Better we force the callers to pass the scope. Thoughts?

**Comment by nija-at** [comment]
> > The _addIntegration could be called from another place outside of the scope of a route. So I thought a default of the API as the parent might be a good idea.
> 
> I am just thinking if this problem would then occur in that context? Better we force the callers to pass the scope. Thoughts?

Let's make `scope` mandatory and make it the first parameter of the method (to keep it similar to our existing patterns). Better to be explicit than implicit. This is an internal API, and we can change it later.

**Review by ayush987goyal — APPROVED** [review]

File: packages/@aws-cdk/aws-apigatewayv2/test/http/route.test.ts [hunk 1]
```diff
@@ -115,6 +115,27 @@ describe('HttpRoute', () => {
     expect(stack2).toCountResources('AWS::ApiGatewayV2::Integration', 1);
   });
 
+  test('route can be defined in sub stack', () => {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/test/http/route.test.ts:118]:
```suggestion
  test('route defined in a separate stack does not create cycles', () => {
```

**Review by nija-at — APPROVED** [review]

**Review by nija-at — APPROVED** [review]

**Pull request merged**